### ISSUE TITLE
Bug: no-unused-vars unused args not reliably detected

### ISSUE THREAD

**Issue by johnnyreilly** [typescript-eslint/typescript-eslint#8498]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=FAUwHgDg9gTgLgAgGYFcB2BjOBLKaEAmIAtlACoAWIACjFAEYA2JAzgBTAILoosgEBBNAQBKIaPH4CYAcwBcCelCjMAhmgA0CAPTbuaXvwTqCCNFEQxxsOP04IWUYiGkxVATwUs4MbGhkA2gC6GsAAlAgA3vYYeI7MAHSMUDJsjs6uHmEA3MAAvsBAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtcNhyJo0DtEiyIAXVkBfEFqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSJsS6ACYAFAIaYN8dHizCwAMw2plUgL4gLQA&tokens=false

### Repro Code

```TypeScript
export function demoTheProblems(
  unusedAndReportedArg: boolean, // unused and not reported by ESLint
  someArray: string[],
) {
  console.log(someArray);
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": [
      "error"
    ]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUnusedParameters": false
  }
}
```

### Expected Result

I expected that `unusedAndReportedArg` should be reported as unused.

### Actual Result

It was not.

### Additional Info

Interestingly, if you tweak the code to what is below:

```ts
export function demoTheProblems(
  unusedAndReportedArg: boolean, // unused and reported by ESLint
  // someArray: string[],
) {
  // console.log(someArray);
}
```

Then the issue is surfaced! See https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=FAUwHgDg9gTgLgAgGYFcB2BjOBLKaEAmIAtlACoAWIACjFAEYA2JAzgBTAILoosgEBBNAQBKIaPH4CYAcwBcCelCjMAhmgA0CAPTbuaXvwTqCCNFEQxxsOP0469LKMRDSYqgJ4KWcGNjQyANoAuhrAAJQIAN72uggYeE7MAHSMUDJsTi5unuEA3MAAvsBAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtcNhyJo0DtEiyIAXVkBfEFqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSJsS6ACYAFAIaYN8dHizCwAMw2plUgL4gLQA&tokens=false

I found this as I was writing up usage of the rule discussed in https://github.com/typescript-eslint/typescript-eslint/issues/8464

**Comment by auvred** [comment]
I think `{ "args": "all" }` should do the trick! [playground](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=FAUwHgDg9gTgLgAgGYFcB2BjOBLKaEAmIAtlACoAWIACjFAEYA2JAzgBTAILoosgEBBNAQBKIaPH4CYAcwBcCelCjMAhmgA0CAPTbuaXvwTqCCNFEQxxsOP04IWUYiGkxVATwUs4MbGhkA2gC6GsAAlAgA3vYYeI7MAHSMUDJsjs6uHmEA3MAAvsBAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgwBtcNhyJo0DtEgAaTJHEBzSVFHx4kAL6yIAXVkmjQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSJsS6ACYAFAIaYN8dHizCwAMw2plUgL4gLQA&tokens=false)

ESLint docs: https://eslint.org/docs/latest/rules/no-unused-vars#args-all

**Comment by johnnyreilly** [comment]
That does seem to work!

**Comment by johnnyreilly** [comment]
Thanks @auvred !
